cmd/utils_test.go

```diff
@@ -0,0 +1,85 @@
+package cmd
+
+import (
+	"testing"
+)
+
+func Test_dfs_simple(t *testing.T) {
+
+	/*
+		Graph:
+				  A
+				/ | \
+			   B  C  D
+				\/   |
+				E	G
+				|
+				F
+				|
+				H
+	*/
+
+	graph := make(map[string][]string)
+	graph["A"] = []string{"B", "C", "D"}
+	graph["B"] = []string{"E"}
+	graph["C"] = []string{"E"}
+	graph["D"] = []string{"G"}
+	graph["E"] = []string{"F"}
+	graph["F"] = []string{"H"}
+	graph["B"] = []string{"E"}
+
+	dp := make(map[string]int)
+	visited := make(map[string]bool)
+	longestPath := make(map[string]string)
+	for k := range graph {
+		if visited[k] == false {
+			dfs(k, graph, dp, visited, longestPath)
+		}
+	}
+	if dp["A"] != 4 {
+		t.Errorf("Max depth of dependencies was incorrect")
+	}
+	if longestPath["A"] != "C" || longestPath["C"] != "E" || longestPath["E"] != "F" || longestPath["F"] != "H" {
+		t.Errorf("Longest path was incorrect")
+	}
+}
+
+func Test_dfs_cycle(t *testing.T) {
+
+	/*
+		Graph:
+					 A
+				   /   \
+				  B     C
+				  |     |
+				  D 	E
+				/   \
+				H	F
+				 \ /
+				  G
+	*/
+
+	graph := make(map[string][]string)
+	graph["A"] = []string{"B", "C"}
+	graph["B"] = []string{"D"}
+	graph["C"] = []string{"E"}
+	graph["D"] = []string{"F"}
+	graph["F"] = []string{"G"}
+	graph["G"] = []string{"H"}
+	graph["H"] = []string{"D"}
+
+	dp := make(map[string]int)
+	visited := make(map[string]bool)
+	longestPath := make(map[string]string)
+	for k := range graph {
+		if visited[k] == false {
+			dfs(k, graph, dp, visited, longestPath)
+		}
+	}
+	if dp["A"]-1 != 5 {
```

Comment:
this would deserve an explanation - why are we substracting one here?

Reply:
For a cycle like this: A -> B -> C -> D -> A
dp["A"] will be 4 since it actually stops after reaching "A" again. But the actual length in this case is 3 (A -> B -> C -> D). I'll add this as a comment above the test for clarity.

I think I should also update the definition of length in the design document so that's it's more clear. 
For A -> B -> C -> D would it be better to define max depth of dependencies as 4 (counting all the vertices) or as 3 (counting all the edges). Most [articles](https://www.geeksforgeeks.org/longest-path-in-a-directed-acyclic-graph-dynamic-programming/) I found had defined it as the number of edges (3 here) but in our case number of vertices (4 here) might make more sense. Let me know your thoughts.
cc @dims

Reply:
max depth of dependencies was defined as the longest possible chain. we could have another name for the other one if you can come up with one :) yes, both would be useful.